core/src/ledger_cleanup_service.rs

```diff
@@ -124,58 +134,63 @@ impl LedgerCleanupService {
+        );
+        *last_purge_slot = root;
 
-            let (num_shreds_to_clean, lowest_slot_to_clean, mut first_slot, total_shreds) =
-                Self::find_slots_to_clean(blockstore, root, max_ledger_shreds);
+        let (slots_to_clean, lowest_cleanup_slot, first_slot, total_shreds) =
+            Self::find_slots_to_clean(&blockstore, root, max_ledger_shreds);
 
-            if num_shreds_to_clean > 0 {
-                debug!(
-                    "cleaning up to: {} shreds: {} first: {}",
-                    lowest_slot_to_clean, num_shreds_to_clean, first_slot
-                );
-                loop {
-                    let current_lowest =
-                        std::cmp::min(lowest_slot_to_clean, first_slot + DEFAULT_PURGE_BATCH_SIZE);
+        if slots_to_clean {
+            info!(
+                "purging data from slots {} to {}",
+                first_slot, lowest_cleanup_slot
+            );
 
+            let purge_complete = Arc::new(AtomicBool::new(false));
+            let blockstore = blockstore.clone();
+            let purge_complete1 = purge_complete.clone();
+            let _t_purge = Builder::new()
+                .name("solana-ledger-purge".to_string())
+                .spawn(move || {
                     let mut slot_update_time = Measure::start("slot_update");
-                    *blockstore.lowest_cleanup_slot.write().unwrap() = current_lowest;
+                    *blockstore.lowest_cleanup_slot.write().unwrap() = lowest_cleanup_slot;
                     slot_update_time.stop();
 
-                    let mut clean_time = Measure::start("ledger_clean");
-                    blockstore.purge_slots(first_slot, Some(current_lowest));
-                    clean_time.stop();
-
-                    debug!(
-                        "ledger purge {} -> {}: {} {}",
-                        first_slot, current_lowest, slot_update_time, clean_time
+                    let mut purge_time = Measure::start("purge_slots_with_delay");
+                    blockstore.purge_slots_with_delay(
+                        first_slot,
+                        lowest_cleanup_slot,
+                        delay_between_purges,
                     );
-                    first_slot += DEFAULT_PURGE_BATCH_SIZE;
-                    if current_lowest == lowest_slot_to_clean {
-                        break;
-                    }
-                    thread::sleep(Duration::from_millis(500));
+                    purge_time.stop();
+                    info!("{}", purge_time);
+                    purge_complete1.store(true, Ordering::Relaxed);
+                })
+                .unwrap();
+
+            // Keep pulling roots off `new_root_receiver` while purging to avoid channel buildup
+            while !purge_complete.load(Ordering::Relaxed) {
+                if let Err(err) = Self::receive_new_roots(new_root_receiver) {
+                    debug!("receive_new_roots: {}", err);
```

Comment:
We use this to exit the thread on receiver disconnect, should we return the error here?

Reply:
The top of `cleanup_ledger()` does:
```rust
        let root = Self::receive_new_roots(new_root_receiver)?;
```
which preserves the previous behaviour of exiting on receiver disconnect.  

The job of the second `receive_new_roots()` here to just wait until compaction is completed to avoid dropping the new cleanup thread before it cleanly returns.  On a receiver disconnect, once compaction completes then `cleanup_ledger()` will be called again by the `solana-ledger-cleanup` thread loop and a thread exit will be triggered.

So I think the answer to your question is, "no"

Reply:
Sure, but since this loop is taking minutes/hours, nobody will be waiting that long for the compaction thread to finish to exit the thread. The cleanup is a bit lazy, but I guess we just ctrl-c everything anyway.

Reply:
It's also no worse than before where the loop still took minutes/hours, without servicing the channel, before returning the error from the `solana-ledger-cleanup`.  
But I'm thinking that we should just remove the compaction entirely.  I'm not convinced it's actually doing anything, and even if it is the 30x disk IO cost is really high